### ISSUE TITLE
Uncaught Error: Invariant Violation: Expected to find a valid target.

### ISSUE THREAD

**Issue by wdhorton** [gaearon/react-dnd#342]
I'm trying to implement the Simple Sortable example in Rails and ES5, and for some reason I keep getting the error message: "Uncaught Error: Invariant Violation: Expected to find a valid target." I looked through the issues and found a few other people who had experienced the same thing, but a solution was never posted. My code is here: https://github.com/wdhorton/dragging_app. The relevant files are in the app/assets/javascripts folder.

**Comment by gaearon** [comment]
I'd like to help but unfortunately I don't have the time to learn Rails ;-)
It's hard to guess what's wrong from the code that looks (almost?) exactly like an example that works.

It would help if you:

* Provided the full error stack trace
* Provided the exact steps to reproduce the issue
* Provided what *exactly* has been changed compared to the example
* Best of all, provided a JavaScript project that can be launched with JavaScript tooling

For now I'll close because it's not actionable for me. But if you provide at least a diff of your changes, a stack trace, and the steps to repro, I'll be happy to reopen.

**Comment by wdhorton** [comment]
C'mon you don't want to learn Rails? Jk, thanks for taking a look at this. This error inspired me to start learning ES6 and start using Javascript tooling for my React projects, so it was productive even though I never really figured out what the original cause was.

Thanks for your work on this, and other projects! I've used React DnD in a few projects now, and I'd love to help out. I'll keep my eye on the issues list.

**Comment by gaearon** [comment]
:+1: Let me know if you find a way to repro!

**Comment by ksonnad** [comment]
I experienced this problem when my components remounted during reorder (due to issues with nested data structures).  Once I fixed the unmounting, the error went away.  You might try adding a componentWillUnmount method to your Card class (that logs to the console or something) so you can determine whether you have the same issue.

**Comment by ericgio** [comment]
@ksonnad: I'm running into a similar issue where all the items in my sortable list are being unmounted (then re-mounted), triggering the ReactDnD error. I'm still trying to figure out what's causing this though. What ended up being the issue for you?

**Comment by ksonnad** [comment]
First check that you're providing the items a key, as described [here](https://facebook.github.io/react/docs/multiple-components.html#dynamic-children).  

The cause of my issue was pretty specific.  I was mapping over nested objects arrays (instead of flat), so the keys were shifting when items moved between one of the nested arrays.

**Comment by jhartma** [comment]
I get the same issue. @ksonnad I'm not sure if I understand you correctly: does unmounting produce the error? And how did you fix it?

**Comment by ksonnad** [comment]
I experienced this error when draggable components unmounted and remounted (rather than just moved) during reordering.  My list of draggable components were rendered from nested arrays that React flattened to produce a single list of children.  However, I think it maintained the hierarchy of the nesting in its representation of the component tree (evidenced by the data-reactid attribute).  So even though the keys I passed to my components were stable, React saw them as nodes of different branches that had to be unmounted and remounted rather than just moved.  When I rendered a pre-flattened array, the unmounting stopped and the issue disappeared.

[Reconciliation | React](https://facebook.github.io/react/docs/reconciliation.html)
[Multiple Children - React](https://facebook.github.io/react/docs/multiple-components.html)

**Comment by ericgio** [comment]
I found the source of my problem, and like in @ksonnad's case, it was React unmounting/re-mounting the children, not a problem with ReactDnD. My issue was that a request was being sent on `endDrag`, and if I tried dragging again before the response came back, React got confused and unmounted all the children. In any case, ReactDnD works fine but because it throws an error (as expected) it creates a somewhat confusing symptom.
